### ISSUE TITLE
Co2L: Contrastive Continual Learning

### ISSUE THREAD

**Issue by chaerinkong** [reyllama/MIPAL-CL#2]
## TL; DR

- Supervised contrastive learning with memory buffers
- Instance-wise relation distillation to prevent catastrophic forgetting

## Motivation

- Apart from combatting catastrophic forgetting, it's important to learn memory-worthy representations (forward transfer) in the first place.

## Formulation

![image](https://user-images.githubusercontent.com/46308407/158129323-a1c6414f-294d-49f4-946d-8fb10bd65865.png)
![image](https://user-images.githubusercontent.com/46308407/158129363-6fdc4fd3-6746-4319-89b3-8e942846b4bb.png)
![image](https://user-images.githubusercontent.com/46308407/158129389-953c62c0-d419-422a-b17a-32f922ae6648.png)

## Evaluation

![image](https://user-images.githubusercontent.com/46308407/158129415-1f4b9387-2d58-40ea-aa04-9fb04306f25a.png)
![image](https://user-images.githubusercontent.com/46308407/158129429-9f2c5ef9-cd1e-490f-bfa3-90c894533f49.png)
![image](https://user-images.githubusercontent.com/46308407/158129460-93529b7c-c599-470c-962f-185df1975352.png)
